### PR TITLE
rustc_metadata: Preprocess search paths for better performance

### PR DESCRIPTION
**Pull request by osiewicz**
Over in Zed we've noticed that loading crates for a large-ish workspace (~100 members of workspace, over 1000 crates being built for the main binary target) can take almost 200ms. We've pinned it down to how rustc searches for paths to dependency files, as it performs a linear search over the list of candidate paths. In our case the candidate list had about 20k entries which we had to iterate over for each dependency being loaded. Our workspace is also pretty bottom-heavy, e.g. most of the workspace members pull in most of the transitive dependencies one way or another, which means that we spend quite some time loading crates at rustc startup.

This commit introduces a simple FilesIndex that's just a BTreeMap under the hood. Since crates are looked up by both prefix and suffix, we perform a range search on said BTree (which constraints the search space based on prefix) and follow up with a linear scan of entries with matching suffixes.

Overall, this commit brings down build time for us in dev scenarios by about 6%. 100ms might not seem like much, but this is a constant cost that each of our workspace crates has to pay, even when said crate is miniscule.

### CODE REVIEW

**Comment by workingjubilee** [comment]
I am pretty sure this should get the attention of

r? @saethlin

File: compiler/rustc_session/src/search_paths.rs [hunk 1]
```diff
@@ -10,9 +13,40 @@ use crate::filesearch::make_target_lib_path;
 pub struct SearchPath {
     pub kind: PathKind,
     pub dir: PathBuf,
-    pub files: Vec<SearchPathFile>,
+    pub files: FilesIndex,
 }
 
+/// [FilesIndex] contains paths that can be efficiently looked up with (prefix, suffix) pairs.
+#[derive(Clone, Debug)]
+pub struct FilesIndex(BTreeMap<Arc<str>, Arc<SearchPathFile>>);
```
Comment (Mark-Simulacrum) [inline compiler/rustc_session/src/search_paths.rs:21]: Since this is built once and then reused, it might make sense to use `Vec<SearchPathFile>` or similar and sorting it + [partition_point](https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.partition_point) for lookup. I'd expect that to be both faster on lookup and might allow skipping the Arcs?

File: compiler/rustc_metadata/src/locator.rs [hunk 2]
```diff
@@ -247,7 +247,7 @@ pub(crate) struct CrateLocator<'a> {
     cfg_version: &'static str,
 
     // Immutable per-search configuration.
-    crate_name: Symbol,
+    pub crate_name: Symbol,
```
Comment (petrochenkov) [inline compiler/rustc_metadata/src/locator.rs:250]: Unnecessary change.

File: compiler/rustc_metadata/src/locator.rs [hunk 3]
```diff
@@ -409,45 +409,40 @@ impl<'a> CrateLocator<'a> {
...
+                            CrateFlavor::Rlib => rlibs.insert(path, search_path.kind),
+                            CrateFlavor::Rmeta => rmetas.insert(path, search_path.kind),
+                            CrateFlavor::Dylib => dylibs.insert(path, search_path.kind),
+                        };
+                    }
+                }
+            }
+            if let Some(static_matches) = spf.query(staticlib_prefix, staticlib_suffix) {
+                for (_, spf) in static_matches {
+                    self.crate_rejections.via_kind.push(CrateMismatch {
```
Comment (petrochenkov) [inline compiler/rustc_metadata/src/locator.rs:441]: This change relies on staticlib prefix/suffix not being the same as any of the rmeta/rlib/dylib prefixes/suffixes, but that seems ok (at least for built-in targets).

Comment (osiewicz) [reply compiler/rustc_metadata/src/locator.rs:441]: That's true; previous code would never fall into static_matches branch as it was an if/else ladder. Fixing this seems as easy as checking whether any of the prefixes match tho.

**Comment by osiewicz** [comment]
In https://github.com/rust-lang/rust/pull/132910/commits/081797d716123886319b23f975d18179f90fe625 I've additionally started filtering out entries that cannot match any query; basically, given a target spec we can remove entries from the map before any query takes place.
This brings down the runtime for me by a bit; the size of refined set in Zed is ~5k entries versus 25k without refining. Unrefined set includes files such as CGUs. We don't need to check them.

File: compiler/rustc_metadata/src/creader.rs [hunk 4]
```diff
@@ -502,17 +502,16 @@ impl<'a, 'tcx> CrateLoader<'a, 'tcx> {
...
         locator.target = &self.sess.host;
         locator.tuple = TargetTuple::from_tuple(config::host_tuple());
-        locator.filesearch = self.sess.host_filesearch(path_kind);
+        locator.filesearch = self.sess.host_filesearch();
+        locator.path_kind = path_kind;
 
         let Some(host_result) = self.load(locator)? else {
             return Ok(None);
         };
-
```
Comment (bjorn3) [inline compiler/rustc_metadata/src/creader.rs:515]: Would you mind preserving the blank lines in this file?

**Comment by saethlin** [comment]
> Overall, this commit brings down build time for us in dev scenarios by about 6%.

Exactly how did you come up with this figure? I've cloned Zed and I'm looking through profiles of builds of the repo and while I certainly can see an impressive amount of time is being spent in this code, I don't think I can come up with an entire 6%. So I'm wondering if I just don't understand how to interact with this repo or if there's some particular workflow I should be investigating.

**Comment by osiewicz** [comment]
@saethlin this figure is an end-to-end timing of `cargo build` in root of zed crate after `touch`ing `crates/editor/src/editor.rs`. This is one of the "hottest" files in our repo and there are multiple crates that depend on editor, thus this is the scenario we've looked at. We use plain `touch` instead of making code changes as we're not familiar with inner workings of incremental build system of rustc; in theory, `touching` should be the best case scenario for incremental build, as nothing really changes (well, macros and whatnot can get in a way, but that's not a factor for us).
I've just reran these measurements against the commit this branch was created from (`fbcdd728ed41a4b5bf314ae7ed18a435a4cc0c34`):
base: 13.37 12.68 12.66 12.75 13.09
test: 11.48 11.59 11.68 11.52 11.81

I did not really mean for this figure to stand as an argument for merging this; it's just that I'm aware that 100ms (or whatever the figure is, depending on a package being built) might not seem like much, but it does add up in the long run for us.

File: compiler/rustc_session/src/search_paths.rs [hunk 5]
```diff
@@ -10,9 +11,52 @@ use crate::filesearch::make_target_lib_path;
...
+    /// Look up [SearchPathFile] by (prefix, suffix) pair.
+    pub fn query<'this, 'prefix, 'suffix>(
+        &'this self,
+        prefix: &'prefix str,
+        suffix: &'suffix str,
+    ) -> Option<impl Iterator<Item = (String, &'this SearchPathFile)> + use<'this, 'prefix, 'suffix>>
+    {
+        let start = self.0.partition_point(|(k, _)| {
+            let k: &str = &k;
+            k < prefix
```
Comment (saethlin) [inline compiler/rustc_session/src/search_paths.rs:31]: This is a bit funky. I'd rather have one of these, but I'd trust you if the form you're using is well-known, just not to my eyes.
```suggestion
            &**k < prefix
```
```suggestion
            **k < *prefix
```

File: compiler/rustc_session/src/search_paths.rs [hunk 6]
```diff
@@ -10,9 +11,52 @@ use crate::filesearch::make_target_lib_path;
...
+        if start == self.0.len() {
+            return None;
+        }
+        let prefixed_items;
+        let end = self.0[start..].partition_point(|(k, _)| k.starts_with(prefix));
+        if end + start == self.0.len() {
+            prefixed_items = &self.0[start..];
+        } else {
+            prefixed_items = &self.0[start..start + end];
+        }
```
Comment (saethlin) [inline compiler/rustc_session/src/search_paths.rs:42]:
```suggestion
        let end = self.0[start..].partition_point(|(k, _)| k.starts_with(prefix));
        let prefixed_items = &self.0[start..][..end];
```

**Comment by saethlin** [comment]
> I did not really mean for this figure to stand as an argument for merging this; it's just that I'm aware that 100ms (or whatever the figure is, depending on a package being built) might not seem like much, but it does add up in the long run for us.

I think you are underestimating how much we care about compiler performance. In general, a 6% win is quite significant.

I don't think I've seen any of the pathology you've identified via Zed in any of our existing benchmark suite. I just checked (what I think is our go-to "big crate"?) cargo-0.60.0 Debug IncrUnchanged (touch a file and run `cargo build`) and it looks completely different from Zed. I've also profiled builds of Zed a fair bit now and I think there are a number of code paths in rustc that are O(n) but scale with either workspace size or dependency tree size. Hopefully some of my other in-progress work will help.

But with just the two style nits, I think this PR is good to go. I'm very happy to see people identify a problem with the compiler and try to fix it :)

**Comment by lqd** [comment]
maybe cleanup the commit history?

**Comment by saethlin** [comment]
My kingdom for bors squash.

@bors r-

@osiewicz can you squash the commits together?
